zcash_client_sqlite/src/wallet/db.rs

```diff
@@ -390,6 +390,25 @@ pub(super) const INDEX_SENT_NOTES_TO_ACCOUNT: &str =
     r#"CREATE INDEX sent_notes_to_account ON "sent_notes" (to_account_id)"#;
 pub(super) const INDEX_SENT_NOTES_TX: &str = r#"CREATE INDEX sent_notes_tx ON "sent_notes" (tx)"#;
 
+/// Stores the set of transaction ids for which the backend required additional data.
+///
+/// ### Columns:
+/// - `txid`: The transaction identifier for the transaction to retrieve state information for.
+/// - `query_type`:
+///     - `0` for raw transaction (enhancement) data,
+///     - `1` for transaction mined-ness information.
+/// - `dependent_transaction_id`: If the transaction data request is searching for information
+///   about transparent inputs to a transaction, this is a reference to that transaction record.
+///   NULL for transactions where the request for enhancement data is based on discovery due
+///   to blockchain scanning.
+pub(super) const TABLE_TX_RETRIEVAL_QUEUE: &str = r#"
+CREATE TABLE tx_retrieval_queue (
+    txid BLOB NOT NULL UNIQUE,
```

Comment:
Thinking through this a bit further, I think that the semantics of `query_type` should be:
* If status is requested, report status.
* If enhancement is requested, call `decrypt_and_store` with the transaction data if available; otherwise, report status (which will then always be `TxidNotRecognized`). 

If a transaction exists in a block that is not in the main chain or is in the mempool, we still want to decrypt and store it, and then it may or may  not eventually be mined or expire. For the purpose of reporting status, we can have an additional `ExpiredUnmined` state for transactions that were mined in an orphaned block or evicted from the mempool but were never mined in the main chain. If a transaction is accepted to the mempool at any point, we should essentially consider it to be in the mempool until it expires, because some other node might have a reference to it in the mempool even if the node our lightwalletd is connected to does not.

Reply:
Let's make sure that `decrypt_and_store` doesn't have undesired side-effects when called for transactions in the mempool (or transactions that don't and possibly can't get mined) before we make that change.

Reply:
This has been resolved as I described above: the `Enhancement` query type now subsumes `Status`, and so the unique constraint on txid is correct.